### PR TITLE
Domain change: Contact form and purchase checkout

### LINKED ISSUE #8407: Domain contact form

### PR DESCRIPTION
**Pull request by 0nko**
Fixes #8407, a subtask of #8238.

First, I want to apologize for the massive PR 🙇. 3/4 of it is the Domain contact from and the associated code, that I copied over from WPAndroid. That's why it's still using XML layout. I wanted to save some time and reuse the existing code since it's been tested and it already contained the transaction logic to purchase a domain using free credits + also things like contact information pre-filling, form validation, etc. The coupling is fairly loose so we can eventually go back and replace contact form with Jetpack Compose, if we want to. 

The domain registration/purchase flow is now as follows:

Domain dashboard screen -> Domain search scree -> `if (hasFreeCredit)` Free domain registration screen (w/ domain contact form) `else` -> Domain purchase screen (webview) -> Purchase success screen

I didn't do much refactoring of the Free domain registration screen besides some minimum changes required to integrate it in the purchase flow. I've thought about splitting the PR somehow but it wouldn't make much difference if I wanted to keep the logical parts of the contact form together. 

If I can suggest a way to review this - ignore the `FreeDomainRegistrationFragment`, its layout XML, `FreeDomainRegistrationViewModel` and `DomainPhoneNumberUtils` at first and leave it for the last as that's the copied code. 

**To test:**

Domain registration using free credit
1. Login using an admin account
2. Select a store that's eligible for a free domain (or create a new one from the app)
3. Go to Settings -> Domains -> Claim a free domain
4. Find a non-premium domain that's free for the first year
5. Tap on Continue
6. If the logged-in WP user has an address stored, it'll be pre-filled
7. If you try to clear any required field like a name, phone number or a country, a validation error is displayed
8. Once the contact details are entered, tap on the Register button
9. Wait for a while until the registration completes
10. Notice the Success screen is shown with a new domain name 

Domain registration via WebView purchase
1. Login using an admin account (give yourself free credits first in SA)
2. Select a store that's _not_ eligible for a free domain 
3. Go to Settings -> Domains
4. Tap on Search for a domain button
5. Select a paid domain
6. Tap on Continue
7. Notice a webview is opened with a checkout website that shows a new domain in a shopping cart
8. Complete the contact information, if it's not filled out already
9. Complete the purchase 
10. Notice the Success screen is shown with a new domain name

### CODE REVIEW

File: WooCommerce/src/main/res/values/themes.xml [hunk 1]
```diff
@@ -144,8 +144,8 @@
     <style name="Woo.Dialog.ButtonBar">
         <item name="android:paddingStart">@dimen/minor_100</item>
         <item name="android:paddingEnd">@dimen/minor_100</item>
-        <item name="android:paddingTop">@dimen/minor_50</item>
-        <item name="android:paddingBottom">@dimen/minor_50</item>
+        <item name="android:paddingTop">@dimen/minor_00</item>
+        <item name="android:paddingBottom">@dimen/minor_00</item>
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/res/values/themes.xml]: hmm why this change? Won't this have unexpected side effects on other Dialogs that use this button bar?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/domainpicker/DomainPickerScreen.kt [hunk 2]
```diff
@@ -202,12 +201,6 @@ private fun DomainSuggestionList(
     keyboardController: SoftwareKeyboardController?,
     modifier: Modifier = Modifier,
 ) {
-    LaunchedEffect(Unit) {
-        suggestions.firstOrNull { it.isSelected }?.domain?.let { domain ->
-            onDomainSuggestionSelected(domain)
-        }
-    }
-
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/domainpicker/DomainPickerScreen.kt:210]: Minor, but I think that by removing this, now when you enter the screen a domain will be preselected for the user but the "Continue" button will be disabled, and the user has to actively click on any domain from the list to get the "Continue" button enabled. 

Is that intentional?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/domain/DomainSuggestionsViewModel.kt [hunk 3]
```diff
@@ -175,7 +176,7 @@ abstract class DomainSuggestionsViewModel constructor(
                         val price = domainSuggestionsRepository.fetchDomainPrice(domain.name).getOrNull()
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/domain/DomainSuggestionsViewModel.kt:183]: Sharing some possible improvement in case you don't have it mind already :)  

I noticed that clicking on different domains from the list when the list contains premium domains is lagging: 

https://user-images.githubusercontent.com/2663464/222405360-888156d2-bdb4-4e9c-a992-961cef1880bf.mp4

I presume this is because every time `onDomainSuggestionSelected` is called then we reach this point of the code where we call `domainSuggestionsRepository.fetchDomainPrice(domain.name)` for each premium domain of the list. A list that hasn't change. The more premium domains the list has, the more lag, and the more unnecessary request we do. 

I wonder if we could have an in memory cache in the repository and save all the prices for the premium domains in a `Map` and only fetch from network if the premium domain is not on the map. Just suggesting this for future improvements. No need to add here.

**Review by JorgeMucientes — APPROVED** [review]

**Comment by JorgeMucientes** [comment]
Excellent job @0nko 🙌🏼 🥇  Handling all the different checkout cases and different domains hasn't been easy for sure. Many thing to keep in mind so :kudos on the fantastic job over all the flow. 

I left some minor np, questions or possible improvements, but nothing blocking.

File: WooCommerce/src/main/res/values/themes.xml (hunk 1 above)
Comment (0nko) [reply WooCommerce/src/main/res/values/themes.xml]: I'm not sure why there was this padding before but it was actually causing the button bar to show scrollbar, which seemed like a bug.

[device-2023-03-02-120815.webm](https://user-images.githubusercontent.com/1522856/222412109-63067aeb-78f0-4e34-b505-848705c8952c.webm)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/domain/DomainSuggestionsViewModel.kt (hunk 3 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/domain/DomainSuggestionsViewModel.kt:183]: Yeah I noticed this as well, but I wanted to finish the feature first to make the code freeze and get back to this later. I'll try to address this when I implement the domain availability. Thanks for the suggestion, though, I'll give it a try!

File: WooCommerce/src/main/res/values/themes.xml (hunk 1 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/res/values/themes.xml]: Oh I see! Thanks for clarifying. What happens if instead of setting the padding to `0dp` we just simply remove those 2 lines? Should be 0dp as well right? Or is there any non 0dp default values I'm missing?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/domainpicker/DomainPickerScreen.kt (hunk 2 above)
Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/storecreation/domainpicker/DomainPickerScreen.kt:210]: Yes, it was intentional because I encountered a situation when it was still causing a crash, so I figured it was just not worth it. The problem was that there wasn't a "single source of truth" for the domain selection and this was a hack, plus it was the "UI" telling the model what the state is. So all wrong 😆. Going back, though, I refactored it a bit and added a [fix](https://github.com/woocommerce/woocommerce-android/pull/8422/commits/68ced7e7c587231c5b675b2ab79a9d67ddf4e323) which should solve the issue. Thanks for pointing it out!

File: WooCommerce/src/main/res/values/themes.xml (hunk 1 above)
Comment (0nko) [reply WooCommerce/src/main/res/values/themes.xml]: Yeah, you're right. Got rid of it 👍

**Comment by 0nko** [comment]
Thanks so much for the review, @JorgeMucientes! I believe I've addressed all of your comments. Ready for another round. 🙂

**Review by JorgeMucientes — APPROVED** [review]

**Comment by JorgeMucientes** [comment]
All looks good thanks for addressing the changes!
